src/docker/files/DockerUri.ts

```diff
@@ -0,0 +1,73 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as corepath from 'path';
+import * as querystring from 'querystring';
+import * as vscode from 'vscode';
+import { DockerOSType } from '../Common';
+
+export type DockerUriFileType = 'directory' | 'file';
+
+export type DockerUriQuery = {
+    containerOS?: DockerOSType;
+    fileType?: DockerUriFileType;
+    serverOS?: DockerOSType;
+};
+
+export class DockerUri {
+
+    public static create(containerId: string, path: string, options?: DockerUriQuery): DockerUri {
+        return new DockerUri(containerId, path, options);
+    }
+
+    public static parse(uri: vscode.Uri): DockerUri {
+        const containerId = uri.authority;
+        const path = uri.path;
+
+        const query = <DockerUriQuery>querystring.decode(uri.query);
+
+        return DockerUri.create(containerId, path, query);
+    }
+
+    public static joinPath(baseUri: DockerUri, ...pathSegments: string[]): DockerUri {
+        const joinedPath = corepath.posix.join(baseUri.path, ...pathSegments);
```

Comment:
Can't use `vscode.Uri.joinPath`?

Reply:
While `DockerUri` is somewhat modeled after `vscode.Uri`, it's not a wrapper (so there's no underlying `Uri` instance). 
 I was attempting to keep the various properties of the Docker path in an object-like form until a `Uri` was actually required (at which point it's constructed on the fly).  To use `vscode.Uri.joinPath()` I'd need to generate the `Uri`, then join it, then parse it back into object form.

Maybe it'd be better to rename the type something like `DockerPath` to avoid the direct comparison to `vscode.Uri`?

Reply:
That makes sense; since the methods are the same though I do like it being `DockerUri`.